### DISCUSSION TITLE
OpenAI, Azure, function calling and Llama CPP grammar integration

### DISCUSSION THREAD

**Discussion by AriMKatz** [outlines-dev/outlines#516] [Feature requests]
Hi, 

I have a few questions about outlines please:

Does outlines  support openAI function calling in general and for json generation? 
Does outlines  utilize llamacpp grammars?
Which features are disabled or don't work well in production for each of  OpenAI and Azure?

Thanks!

**Comment by brandonwillard** [comment]
> Does outlines utilize llamacpp grammars?

No, `outlines` has a distinct approach to guided generation and does not host `llama.cpp`'s.  For regular grammars, the `outlines` approach is near/at a theoretical optimum for run-time performance, so I don't think we can reasonably consider using or offering anything else, especially since the more general context-free grammar case is being expanded to have similar performance as well.
